### PR TITLE
feat(record): structural verify guards for the produced video

### PR DESCRIPTION
**Pull request by hawkeyexl**
## Summary

Layer 4 (final) of the recording visual-regression plan: **structural `verify` guards on the produced video** ([ADR 01080](https://github.com/doc-detective/doc-detective/blob/claude/recording-verify/adrs/01080-structural-recording-assertions-via-ffmpeg-probes.md)). Layers 1-3 (#645, #649, #651) have landed; this now targets `main` directly.

Checkpoints validate the *scene* a recording shows; `verify` validates the *file*. It catches the historical failure modes of screen recording — an all-black capture, a crop landing outside the window, a truncated clip — which today pass silently because the steps themselves didn't throw.

```json
{
  "record": {
    "path": "signup-flow.mp4",
    "verify": { "minDuration": 2, "resolution": true, "notBlack": true }
  }
}
```

- `minDuration` / `maxDuration` — the video's length.
- `resolution` — `true` compares against the capture plan's resolved expectation (the crop rectangle for `window`/`viewport` targets, else the capture frame size; skipped when the engine has no expectation), ±2 px for even-dimension rounding. The object form compares literals.
- `notBlack` — a bounded ffmpeg `blackdetect` pass.

Unlike checkpoint drift (advice → WARNING), a violated guard **FAILs** the step: these are properties the author explicitly demanded. Verify and checkpoint specs evaluate in one implicit-assertion pass, so a structural FAIL outranks drift's WARNING. Measured values land in outputs (`$$resolutionMatch`, `$$allBlack`).

## The interesting bug (found by running it, not reading it)

The first `notBlack` implementation reported a fully-black clip as *not* black. Probing real ffmpeg showed why: **blackdetect's final interval ends at the last black frame's timestamp, not the clip end** — a 0.5 s / 10 fps black clip reports `black_duration:0.4`, an under-report of exactly one frame interval that a 95% coverage threshold can't absorb on short clips. The probed `fps` now sizes that tolerance; without an fps the shortfall is indistinguishable from real content, so the check stays undecided rather than guessing. Both the signature and the half-black counter-case are pinned by tests.

**No verdict without evidence** (self-review): `notBlack` skips when the duration is unknown rather than reporting `allBlack: false` it had no basis for. Duration guards still fail *closed* — an author who demanded a floor shouldn't get a silent pass on an unprobeable file.

## Fixtures

The headed mp4 leg of `recording-outputs.spec.json` gains a passing `verify` block — `minDuration`, `maxDuration`, `resolution: true`, `notBlack` — so every guard runs end-to-end through the real probe at least once.

Two permutations can't be fixtures, and the ADR now says so rather than leaving a silent gap: a violated guard would FAIL (fixtures never do), and `resolution`'s object form can't be deterministic when the capture is the runner's display, whose size varies by image. Both stay in mocha.

## Docs impact

Yes — new action option. `record.mdx` gains "Verifying the video file itself"; `stopRecord.mdx` documents the new outputs; schema references regenerated.

## Verification

- 18 layer-4 unit tests green, including `detectAllBlack` against real ffmpeg-generated black and red clips; 427 recording-related tests green; full suite green; src/common coverage ratchet passes.
- End-to-end through the real CLI (headed Windows), re-verified after the merge: spec **PASS** with all four guards evaluated against a real file — `duration >= 0.5`, `duration <= 60`, `resolutionMatch == true`, `allBlack == false`, against probed `duration=2.1`, `resolutionMatch=true`, `allBlack=false`.

### Note on `src/core/tests.ts`

The one-line `handle.verify` change there is genuinely one line. An earlier commit in this branch flipped the file's CRLF endings into a 5,878-line phantom diff (that file is CRLF-committed while most are LF, and `core.autocrlf=input` re-strips CRs on `git add`); it's restored, and the landmine is documented in `CLAUDE.md`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> **ADR numbering note:** #644 claimed 01071-01073 while this stack was open, and #655/#656 merged ahead of it taking 01076/01077. This ADR therefore lands as **01080**, leaving `main` gapless at 01070-01080 with no duplicates.
>
> Merging `main` in also required dropping this branch's own snapshots of the layer 1-3 ADRs: it predated the stack landing, so it carried them under their pre-merge numbers (01071-01074) while `main` has all four under their final numbers *with* the review fixes these copies never got. Git's rename detection surfaced that as delete/modify conflicts; resolved toward `main` in every case.
>
> Citations were re-pointed with care rather than sed'd: verify is 01080, the stopRecord probe is 01074, checkpoints is 01075. Two of those were already stale on `main` and are corrected here. The `ADR 01072` references in `tests.ts` and `utils.ts` are deliberately untouched — they cite the bidi-socket decision, which really is 01072. Same number, different ADR.

## Summary by CodeRabbit

- **New Features**
  - Added opt-in `record.verify` structural checks for minimum/maximum duration, resolution match, and non-black content.
  - Verification runs when recording stops and can fail the `stopRecord` step; structural failures take precedence over checkpoint drift.
  - When applicable, `stopRecord` now reports `resolutionMatch` and `allBlack` verdicts.
- **Documentation**
  - Documented post-record “video file” verification in `record.mdx`/`stopRecord.mdx`, including updated guard outputs and failure behavior.
  - Added schema references for `verify` and related schemas.
- **Tests**
  - Added schema validation and end-to-end verification tests, including black-content detection and stream-parsing edge cases.
- **Chores**
  - Updated contributor guidance and CI flake-triage notes.

### CODE REVIEW

File: docs/fern/pages/docs/actions/record.mdx [hunk 1]
```diff
@@ -200,6 +200,31 @@ Checkpoints detect drift; `overwrite: "aboveVariation"` acts on it. With this se
 
 `aboveVariation` requires checkpoints, so it enables them automatically—including over an explicit `checkpoints: false`, which this mode overrides (a verdict with no evidence could never refresh a drifted recording). Add a `checkpoints` object to tune them. When the recording refreshes, its baselines update with it—the two always stay in sync—and the `stopRecord` step reports `changed: true` with the reasons in its outputs. If a step fails during the span, Doc Detective keeps the existing recording and baselines untouched rather than judging drift from an incomplete run.
 
+### Verifying the video file itself
+
+Checkpoints validate the scene a recording shows; `verify` validates the produced file. Use it to assert structural properties of the video—the classic capture failure modes are an all-black video, a wrong-sized crop, or a truncated clip:
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply docs/fern/pages/docs/actions/record.mdx]: Fixed — recast as a clause instead of a semicolon: "Checkpoints validate the scene a recording shows, while `verify` validates the produced file." Reads better anyway.

File: src/core/tests/stopRecording.ts [hunk 2]
```diff
@@ -618,34 +623,125 @@ async function stopRecording({
...
     result.assertions = assertions;
     result.status = status;
-    if (status === "WARNING") {
+    // Only verify guards carry FAIL severity; checkpoint specs are warnings.
+    // So a FAIL here is always a guard, and a WARNING always has checkpoints —
+    // the `entries` check keeps that assumption honest rather than implicit,
+    // since the guards can run on a span with no checkpoints at all.
+    if (status === "FAIL") {
+      result.description += ` One or more structural verify guards failed.`;
+    } else if (status === "WARNING" && checkpoints?.entries?.length) {
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply src/core/tests/stopRecording.ts:744]: Fixed — and this one was real, thank you. Neither PR had it alone: `aboveVariation` (ADR 01078) and `verify` (ADR 01080) only meet on the same span after the merge, which is what made the downgrade reachable.

The step didn't just report the wrong status, it contradicted itself. Red test output before the fix:

```
got WARNING: Stopped recording. One or more structural verify guards failed.
The captured video couldn't replace ...\out.mp4, so the existing (now stale)
recording was kept.
```

It says a structural guard failed, and reports WARNING. Any gate keyed on FAIL waves that straight through.

Now `if (result.status !== "FAIL") result.status = "WARNING";` — FAIL survives, and the stale-target note still appends, because that's additive information rather than a verdict.

Pinned by an integration test that forces a genuine promote failure rather than stubbing one: a non-empty `<target>.promote-backup` directory defeats both the non-recursive `rmSync` that clears a stale backup and the subsequent rename onto it. The retained target is an all-black clip, so the probe measures it and `notBlack` FAILs. Verified red (`got WARNING`) → green.

File: src/core/tests/ffmpegRecorder.ts [hunk 3]
```diff
@@ -582,6 +584,110 @@ async function probeVideoMetadata({
...
+          const text = chunk.toString();
+          if (text.includes("black_") && collected.length < 65536) {
+            collected = (collected + text).slice(0, 65536);
+          }
+        });
+        proc.on("error", () => done(null));
+        proc.on("close", (code: number | null) =>
+          done(code === 0 ? collected : null)
+        );
+        setTimeout(() => done(null), 30000);
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply src/core/tests/ffmpegRecorder.ts]: Fixed — this was the most dangerous of the three, because it fails in the unsafe direction: an all-black video would have **passed** `notBlack`.

You're exactly right about the mechanism. Chunk boundaries fall wherever the pipe flushes, so `black_duration:0.4` can straddle two chunks and then *neither* contains `black_` — the interval is dropped, coverage reads as 0, and the guard reports "not black". Worse, it's invisible where we test it (the suite's clips are 0.5s and arrive in one chunk) and live exactly where the guard earns its keep: long recordings with enough decode progress spam to force multiple flushes.

Filtering now happens per line, in a pure `createMatchingLineCollector` helper (per the repo's "parsing goes in a small pure helper" rule), which buffers the trailing partial line and flushes any final unterminated one at close — ffmpeg's last write may not end in a newline.

Tests pin the torn cases specifically, since a naive test wouldn't reproduce it:
- a line cut mid-token (`black_st` | `art:0 …`)
- the same line pushed **one character at a time**
- a final line with no trailing newline
- CRLF, and progress spam that must not leak
- bounded growth that still keeps the earlier match
- and the end-to-end shape: a torn line still parses as full coverage → `parseBlackdetect(...) === true`

File: src/core/tests/ffmpegRecorder.ts [hunk 4]
```diff
@@ -582,6 +584,110 @@ async function probeVideoMetadata({
...
+          // long decode must not evict them.
+          const text = chunk.toString();
+          if (text.includes("black_") && collected.length < 65536) {
+            collected = (collected + text).slice(0, 65536);
+          }
+        });
+        proc.on("error", () => done(null));
+        proc.on("close", (code: number | null) =>
+          done(code === 0 ? collected : null)
+        );
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply src/core/tests/ffmpegRecorder.ts]: Fixed. `detectAllBlack` now holds the handle and `done` clears it — matching `probeVideoMetadata`, which already did. `done` is already idempotent via `settled`, so the timer path and the close/error paths can't double-resolve.

File: src/common/test/validate.test.js [hunk 5]
```diff
@@ -410,6 +410,53 @@ import { validate, transformToSchemaKey } from "../dist/validate.js";
...
+      it("should reject a record step with invalid verify guards", function () {
+        // Note: primitives coerce under ajv coerceTypes, so negatives use
+        // out-of-range numbers, unknown keys, incomplete objects, and arrays.
+        for (const verify of [
+          { minDuration: -1 },
+          { unknownGuard: true },
+          { resolution: { width: 1280 } },
+          { resolution: { width: 0, height: 480 } },
+          [true],
+        ]) {
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply src/common/test/validate.test.js:451]: Fixed. Added negatives for both, and verified they genuinely reject rather than assuming (AJV runs with `coerceTypes`, which quietly rescues a lot of would-be negatives):

- `{ maxDuration: -1 }` — below `minimum: 0`
- `{ minDuration: "soon" }` — uncoercible to number
- `{ notBlack: {} }` — an object won't coerce to boolean

All three fail validation as intended.

File: test/core-artifacts/recording/recording-outputs.spec.json [hunk 6]
```diff
@@ -28,7 +28,13 @@
...
               "target": "display"
             },
-            "overwrite": "true"
+            "overwrite": "true",
+            "verify": {
+              "minDuration": 0.5,
+              "maxDuration": 60,
+              "resolution": true,
+              "notBlack": true
+            }
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply test/core-artifacts/recording/recording-outputs.spec.json:37]: Partially fixed — the two asks that can be deterministic are in; the rest can't be fixtures, and the ADR now says why rather than leaving a silent gap.

**Done — asserting the guards actually measured.** This was the strongest part of your point. The fixture captured the metadata outputs but never `resolutionMatch`/`allBlack`, so a guard that silently stopped running would have left the spec green while proving nothing. Now:

```
resolutionMatch === 'true' && allBlack === 'false'
```

Verified headed: it reports `duration=2.1`, `resolutionMatch=true`, `allBlack=false`, and all four guards appear as assertion records (`duration >= 0.5`, `duration <= 60`, `resolutionMatch == true`, `allBlack == false`).

**Done — the disabling forms.** New `recording-verify-disabling-forms` test: `resolution: false` + `notBlack: false` with a live `minDuration` alongside. It asserts the duration guard still ran *and* that the disabled guards reported **no verdict at all** — their outputs stay unresolved rather than defaulting to a pass the run never made. (Confirmed against the code: `stopRecording.ts:648` skips on `resolution: false`, `:687` on `notBlack: false`, and both outputs are only set inside those branches.) The other disabling form — omitting `verify` entirely — is the gif test in the same spec. `maxDuration` was added in the previous commit.

**Can't be fixtures, now documented in ADR 01080:**
- **A violated guard** would resolve the spec to FAIL, and the gate fails on any FAILed spec. Stays in mocha.
- **`resolution`'s object form** compares literal dimensions, but the capture is the runner's display — 3840×2160 here, something else on hosted macOS/Windows images. There's no deterministic literal to assert. Stays in mocha, where the match/mismatch pair is already covered.
- **The skip paths** (`resolution: true` with no capture expectation; `notBlack` with unknown duration) need a device-engine handle or an unprobeable file — both are unit-level by construction. Already covered there.

Full spec verified headed on Windows: PASS across all three tests.

File: docs/fern/pages/reference/schemas/record-detailed.md [hunk 7]
```diff
@@ -20,6 +20,7 @@ directory | string | Optional. Directory of the file. If the directory doesn't e
...
 name | string | Optional. Identifier for this recording. A later `stopRecord` step can target it by name (`stopRecord: "<name>"`), which is how you stop a specific recording when several overlap. Names must be unique among recordings that are active at the same time. If omitted, the recording is anonymous and is stopped LIFO by an untargeted `stopRecord`.<br/><br/>Pattern: `\S` | 
 engine | one of:<br/>- string<br/>- object([Recording engine (detailed)](/reference/schemas/recording-engine-detailed)) | Optional. Recording engine to use. Either a string shorthand selecting the engine with defaults, or an object for full control. If unset, defaults to the `browser` engine when a visible Chrome context is available and to `ffmpeg` otherwise. | 
+verify | object([Recording verify guards](/reference/schemas/recording-verify-guards)) | Optional. Structural assertions on the produced video file, evaluated when the recording stops. Unlike checkpoint drift (a WARNING), a violated guard FAILs the stopRecord step — these are properties you explicitly demand of the artifact. |
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply docs/fern/pages/reference/schemas/record-detailed.md:23]: Skipping — this file is generated, and the pattern is repo-wide rather than anything this PR introduced.

`docs/fern/pages/reference/schemas/record-detailed.md` is emitted by `docs/.scripts/buildSchemaReferencesV4.js` (`npm run docs:build-schema-refs`). Every property without a `default` gets an empty final cell, so the same row shape exists in **84** of the generated schema reference pages on `main`:

```
$ grep -l ' | $' docs/fern/pages/reference/schemas/*.md | wc -l
84
```

The `verify` row just happens to be a new instance of it. Fixing it properly means changing the generator's Default-column emission (an explicit `—` or similar) and regenerating all 84 pages — a docs-tooling change with its own diff, not something to bury in a recording-feature PR. No CI gate flags it today either; this PR is green.

Filed as a follow-up rather than dropped.

File: docs/fern/pages/reference/schemas/recording-verify-guards.md [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+- [Record (detailed)](/reference/schemas/record-detailed)
+
+## Fields
+
+Field | Type | Description | Default
+:-- | :-- | :-- | :--
+minDuration | number | Optional. Fail if the video is shorter than this many seconds.<br/><br/>Minimum: 0 | 
+maxDuration | number | Optional. Fail if the video is longer than this many seconds.<br/><br/>Minimum: 0 | 
+resolution | one of:<br/>- boolean<br/>- object([Verify resolution (detailed)](/reference/schemas/verify-resolution-detailed)) | Optional. Verify the video's dimensions (±2 pixels — encoders round to even dimensions). `true` compares against the capture plan's resolved expectation (the crop rectangle when a window/viewport crop applied, else the capture frame size; skipped when no expectation exists for the engine). An object compares literal dimensions. | 
+notBlack | boolean | Optional. Fail if the video is essentially all black (black frames cover 95% or more of its duration) — the classic symptom of a broken screen capture. |
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply docs/fern/pages/reference/schemas/recording-verify-guards.md:22]: Skipping — this file is generated, and the pattern is repo-wide rather than anything this PR introduced.

`docs/fern/pages/reference/schemas/record-detailed.md` is emitted by `docs/.scripts/buildSchemaReferencesV4.js` (`npm run docs:build-schema-refs`). Every property without a `default` gets an empty final cell, so the same row shape exists in **84** of the generated schema reference pages on `main`:

```
$ grep -l ' | $' docs/fern/pages/reference/schemas/*.md | wc -l
84
```

The `verify` row just happens to be a new instance of it. Fixing it properly means changing the generator's Default-column emission (an explicit `—` or similar) and regenerating all 84 pages — a docs-tooling change with its own diff, not something to bury in a recording-feature PR. No CI gate flags it today either; this PR is green.

Filed as a follow-up rather than dropped.

File: docs/fern/pages/reference/schemas/recording-verify-guards.md [hunk 9]
```diff
@@ -0,0 +1,33 @@
+---
+title: "Recording verify guards"
+description: "Structural assertions on the produced video file, evaluated when the recording stops. Unlike checkpoint drift (a WARNING), a violated guard FAILs the stopRecord step — these are properties you explicitly demand of the artifact."
+---
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply docs/fern/pages/reference/schemas/recording-verify-guards.md:4]: Skipping here — you're right about the rule, but these files are generated, and fixing them per-file would make things worse.

`docs/AGENTS.md` does say "DO NOT: Use the `.md` file extension" for pages under `fern/pages`. But both of these are emitted by `docs/.scripts/buildSchemaReferencesV4.js`, which hardcodes `.md` in three places (~615, ~636, and the stale-file cleanup at ~649). Every schema reference page is `.md` today:

```
$ ls docs/fern/pages/reference/schemas/ | sed 's/.*\.//' | sort | uniq -c
     93 md
```

So renaming just these two by hand would be undone by the next `npm run docs:build-schema-refs` — it would re-emit the `.md`, and my hand-made `.mdx` would linger as an orphan the cleanup (keyed on `.md`) wouldn't catch. That's strictly worse than the status quo.

The real fix is one line in the generator plus a 93-file regeneration, and it has to be checked against the nav in `docs.yml`, the generator's own cross-links, and the `reference/schemas/...` hrefs in `src/hints/`. That's a docs-tooling change with its own reviewable diff, not something to bury in a recording-feature PR — and it's the same generator as your Default-column comment, so the two should land together.

Filed as a follow-up covering both, rather than dropped.

File: src/core/tests/ffmpegRecorder.ts [hunk 10]
```diff
@@ -582,6 +585,147 @@ async function probeVideoMetadata({
...
+}: {
+  cacheDir?: string;
+  filePath: string;
+  duration?: number;
+  fps?: number;
+}): Promise<boolean | null> {
+  try {
+    const ffmpegPath = await getFfmpegPath({ cacheDir, autoInstall: false });
+    const stderr = await new Promise<string | null>((resolve) => {
+      let collected = "";
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply src/core/tests/ffmpegRecorder.ts]: Fixed. Leftover from the refactor: `createMatchingLineCollector` owns the accumulation now, so `detectAllBlack`'s own `collected` was left declared and never read. Removed — behavior unchanged, 134 recording/ffmpeg tests still green.

**Comment by hawkeyexl** [comment]
All three addressed — the blocking one and the "nit" that turned out to be a real bug.

**Blocking — timer leak in `detectAllBlack`** ✅ Fixed in `3db1a842`. `let timer: NodeJS.Timeout | undefined` alongside `proc`, cleared in `done`, exactly as `probeVideoMetadata` already did. `done` is idempotent via `settled`, so the timeout and close/error paths still can't double-resolve. You were right that it was flagged before and slipped through the round — my apologies for the re-report.

**Nit — the `text.includes("black_")` chunk filter** ✅ Fixed, and I'd push back on "nit": it's a correctness bug that fails in the unsafe direction. If a chunk boundary splits a blackdetect line, *neither* chunk contains `black_`, the interval is dropped, coverage reads 0, and an **all-black video passes `notBlack`** — the exact thing the guard exists to catch. It's invisible in the tests (0.5s clips arrive in one chunk) and lives on long recordings, where flushes are guaranteed. Filtering now happens per line in `createMatchingLineCollector`, pinned by lines torn mid-token and one character at a time.

**Important — `resolution: true` with an `appliedCrop` untested** ✅ Fixed in `10f4fd3c`. You were right that nothing covered it: only the ffmpeg engine sets a crop, and every verify test used a device handle. Two ffmpeg handles with a real crop transcode:

- crop 32×32 of a 64×64 capture → the guard measures the **crop** (PASS, `width`/`height` 32). Had it used `captureInfo.frameSize`, it'd miss by 32px and FAIL — so the test discriminates rather than just passing.
- a 200×200 crop the transcode clamps to the 64×64 source → the produced video can't match what the rect claimed → FAIL. That's the real bug shape: a bad `cropRect`, or a pending-scale factor that overshot.

**Nit — `let proc: any = null`** — skipping. `probeVideoMetadata` uses the same `any`, so tightening one and not the other trades an inconsistency for a cosmetic gain; worth doing across the file in its own change if it's wanted.

Full suite green: 3760 + 974 passing.
